### ISSUE TITLE
Explanation for "#Params" value of 7 on ogbg-molhiv submission ( https://github.com/sangyx/gtrick/blame/a87f225ae851d1fb0c4330712c731723ebf95332/benchmark/pyg/README.md#L14  )

### ISSUE THREAD

**Issue by willy-b** [sangyx/gtrick#10]
Hello @sangyx !
Thanks very  much for sharing this work. 

**For the ogbg-molhiv entry at https://web.archive.org/web/20240324173558/https://ogb.stanford.edu/docs/leader_graphprop/ for the GIN entry which links back to this repository (#24 , GIN, [Yunxin Sang(SJTU)](mailto:sangyunxin@gmail.com)) the "#Params" column has a value of 7.**

![7parameter_GIN_model_on_ogbg_molhiv_does_not_seem_consistent_with_GIN_parameter_counts](https://github.com/sangyx/gtrick/assets/6969859/a6c25ff3-e655-42f8-931f-24df43857133)

**Can you confirm that value is correct, and if so, elaborate how the GIN model is run with only 7 parameters?** 

It appears per https://github.com/sangyx/gtrick/blame/a87f225ae851d1fb0c4330712c731723ebf95332/benchmark/pyg/README.md#L7 that the command would be `python graph_pred.py --model gin` to reproduce the ogbg-molhiv benchmark score.

Specifically, can you confirm whether it was EGNN with EGINConv layers ( https://github.com/sangyx/gtrick/blob/4491cc30f0c8810d8d2cf834fc5f5a8e84e0efc9/benchmark/pyg/model.py#L124 ) which appear to default to 5 layers with 300 channels ( https://github.com/sangyx/gtrick/blame/4491cc30f0c8810d8d2cf834fc5f5a8e84e0efc9/benchmark/pyg/graph_pred.py#L152C7-L153C68 ) now and 256 channels originally ( https://github.com/sangyx/gtrick/blob/108e956a6f7ecbf12202caecf7c7bb4f82757ab3/benchmark/graph_pred.py#L140 )
used for ogbg-molhiv 0.7778 ROCAUC as listed in https://github.com/sangyx/gtrick/blame/4491cc30f0c8810d8d2cf834fc5f5a8e84e0efc9/benchmark/pyg/README.md#L24 , and how many parameters per `sum(p.numel() for p in model.parameters())` this has or had?

Even for a very simple model type like logistic regression, 7 parameters seems to be fewer than the number of atom and bond dimensions in the dataset, so would seem to require looking at only a subset of the dimensions to achieve that number. But for a GIN with multiple layers and nontrivial hidden dimension (regardless of whether 256 or 300 or other changes in repo since initial commit), how this model would only have 7 parameters seems unclear. Weihua Hu (OGB team)'s first demonstration GIN on the leaderboard for this problem had 1,885,206 parameters, by comparison. I do believe it could be done with orders of magnitude fewer, but 7 seems extreme for a GIN.

Thanks in advance for explaining this!

**Comment by sangyx** [comment]
Thanks for your feedback, @willy-b.
You're right: the #Params on ogbg-molhiv are actually #hyperparams. If possible, could you help modify it?

**Comment by willy-b** [comment]
tl;dr: only modification needed to get the current parameter count is **add a print of OGB's recommended expression `sum(p.numel() for p in model.parameters())`, inserted right before `run_graph_pred(args, model, dataset)` at https://github.com/sangyx/gtrick/blob/4491cc30f0c8810d8d2cf834fc5f5a8e84e0efc9/benchmark/pyg/graph_pred.py#L182 to get these values (specifically, you can add `print(f"model parameter count: {sum(p.numel() for p in model.parameters())}")`).**

**For me to help get parameter counts for you, which I am happy to do or double check, I would need to confirm the exact configuration you used in your submission (e.g. whether hidden_channels was 300 or 256) as I have not yet reproduced your result. (I do try to guess below.)**

For reference, per https://web.archive.org/web/20240324175343/https://ogb.stanford.edu/docs/leader_overview/ , it is the parameters of the model (not hyperparameters), to be listed:
```
#Params: The number of parameters of your model, 
which can be calculated by `sum(p.numel() for p in model.parameters())`. 
If you use multi-stage training (e.g., apply node2vec and then MLP), 
please sum up all the parameters (both node2vec and MLP parameters).
```

They also say:
```
Please [contact us](mailto:ogb@cs.stanford.edu) if there is anything unclear or you have found any problematic submission on the leaderboards.
```
If you prefer (were busy or something) I can contact them once we agree on an update to the parameter count to get it updated but probably it would be smoothest if you contacted them there to update it. I am not a member of the OGB team (just took Stanford XCS224W whose professor's group helps manage OGB and which had a homework assignment related to ogbg-molhiv is why I was reviewing the entries) but would be happy to help you get this number corrected.

Using your default configuration ( https://github.com/sangyx/gtrick/blame/4491cc30f0c8810d8d2cf834fc5f5a8e84e0efc9/benchmark/pyg/graph_pred.py#L152C7-L153C68 ), it appears (shown later in this comment) 
**that applying their provided parameter counting expression to your unofficial implementation of the GIN model causes it to have more similar parameter counts to the OGB team's official GIN model entry for ogbg-molhiv, where theirs has 1,885,206 parameters listed in the `#Params` field (entry #33 at https://web.archive.org/web/20240324173558/https://ogb.stanford.edu/docs/leader_graphprop/#ogbg-molhiv )** (without virtual node or other tricks). Makes sense as OGB team also uses 5 layers, 300 emb dim for their GNN https://github.com/snap-stanford/ogb/blob/f631af76359c9687b2fe60905557bbb241916258/examples/graphproppred/mol/main_pyg.py#L72 by default same as default this project has at https://github.com/sangyx/gtrick/blame/4491cc30f0c8810d8d2cf834fc5f5a8e84e0efc9/benchmark/pyg/graph_pred.py#L152C7-L153C68 ). Note it is possible to reduce the parameter count potentially without reducing the performance by reducing the depth (e.g. from 5 to 2, there is evidence from GCNs, see discussion in https://arxiv.org/pdf/1806.03536 that two hops may be enough for many networks) and/or reducing the embedding dimension, but that is a separate discussion from what you are doing here.

![7parameter_GIN_model_on_ogbg_molhiv_may_be_mistake_in_parameter_count_compare_to_ogb_team_GIN-1 8M_parameters_though_can_be_low_thousands_or_smaller_if_number_of_layers_and_embedding_is_made_tiny](https://github.com/user-attachments/assets/e8cc1907-9a2a-4d83-a308-5bc66907728e)

**To get some idea of the numbers expected,**
I tried printing this for the default configuration of your graph_pred.py with GIN for ogbg-molhiv:

For graph_pred.py GIN, ogbg-molhiv with num_layers=5, hidden_channels=300 (your current default; https://github.com/sangyx/gtrick/blame/4491cc30f0c8810d8d2cf834fc5f5a8e84e0efc9/benchmark/pyg/graph_pred.py#L152C7-L153C68 ),
**the number of parameters appears to be 1,885,506 (close to the value reported by the OGB team for their own GIN model**; may make sense as OGB team also uses 5 layers, 
300 emb dim for their GNN https://github.com/snap-stanford/ogb/blob/f631af76359c9687b2fe60905557bbb241916258/examples/graphproppred/mol/main_pyg.py#L72 by default):
```
# in google colab jupyter notebook
%cd /content/
!git clone https://github.com/sangyx/gtrick.git
%cd /content/gtrick/benchmark/pyg
!python graph_pred.py  --model gin --dataset ogbg-molhiv --device 0 --dataset_path /content/dataset # provided this dataset path is an empty writable folder

#Namespace(dataset='ogbg-molhiv', dataset_path='/content/dataset', device=0, log_steps=1, num_layers=5, 
#hidden_channels=300, 
#dropout=0.5, lr=0.001, 
#batch_size=32, 
#num_workers=0, 
#model='gin', epochs=500, runs=10, patience=30)

#model parameter count: 1885506

# printed model object as well
#model:EGNN(
#  (node_encoder): AtomEncoder(
#    (atom_embedding_list): ModuleList(
#      (0): Embedding(119, 300)
#      (1): Embedding(5, 300)
#      (2-3): 2 x Embedding(12, 300)
#      (4): Embedding(10, 300)
#      (5-6): 2 x Embedding(6, 300)
#      (7-8): 2 x Embedding(2, 300)
#    )
#  )
#  (convs): ModuleList(
#    (0-4): 5 x EGINConv() # each has 366001 parameters, not tied/shared
#  )
#  (bns): ModuleList(
#    (0-4): 5 x BatchNorm1d(300, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
#  )
#  (out): Linear(in_features=300, out_features=1, bias=True)
#)
#model.convs[0-4]: EGINConv(), parameters: 366001 per layer
#model.convs[0-4].edge_encoder: BondEncoder(
#  (bond_embedding_list): ModuleList(
#    (0): Embedding(5, 300)
#    (1): Embedding(6, 300)
#    (2): Embedding(2, 300)
#  )
#)
#model.convs[0-4].mlp: Sequential(
#  (0): Linear(in_features=300, out_features=600, bias=True)
#  (1): BatchNorm1d(600, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
#  (2): ReLU()
#  (3): Linear(in_features=600, out_features=300, bias=True)
#)

# This may not be the configuration used by @sangyx as for 10 runs output from author's code on Google Colab L4 GPU, 
# in initial checks I get 0.75 ROCAUC 
# (a bit far from 0.7778 ROCAUC for ogbg-molhiv reported on the leaderboard and 
# https://github.com/sangyx/gtrick/blame/4491cc30f0c8810d8d2cf834fc5f5a8e84e0efc9/benchmark/pyg/README.md#L24 )
# --------------------------------------------------------------------------------
# All runs:
# Best Train: 0.9240 ± 0.0208
# Best Valid: 0.8325 ± 0.0103
# Final Train: 0.8765 ± 0.0300
# Final Test: 0.7505 ± 0.0144
# --------------------------------------------------------------------------------
```

Earlier it appeared graph_pred.py had been configured for a default of num_layers=5, **hidden_channels=256,** ( https://github.com/sangyx/gtrick/blob/108e956a6f7ecbf12202caecf7c7bb4f82757ab3/benchmark/graph_pred.py#L140 ) 
so I checked the parameter count there as well and it appears to **use 1,383,686 parameters in that case:**
```
# in google colab jupyter notebook
%cd /content/
!git clone https://github.com/sangyx/gtrick.git
%cd /content/gtrick/benchmark/pyg
!python graph_pred.py --hidden_channels 256 --model gin --dataset ogbg-molhiv --device 0 --dataset_path /content/dataset

# Namespace(dataset='ogbg-molhiv', dataset_path='/content/dataset', device=0, log_steps=1, 
#num_layers=5, 
#hidden_channels=256, 
#dropout=0.5, 
#lr=0.001, 
#batch_size=32, 
#num_workers=0, 
#model='gin', 
#epochs=500, 
#runs=3, 
#patience=30)

# model parameter count: 1383686
# (performance measured in ROCAUC pending, takes long time to generate 10 runs; please let me know what configuration you ran so I can count parameters and reproduce the score you submitted to ogbg-molhiv for that configuration, thanks!)
```

Thanks!

**Comment by sangyx** [comment]
Hey @willy-b, I've been quite busy lately. If you're interested, you could update the GIN record on OGB with the correct #params and your reproduction performance (And remember to add your name to this record). Feel free to experiment with different hyperparameters (lr, hidden_channels, dropout) and report the best results. The record I'm submitting to the leaderboard aims to demonstrate that even a simple model with deliberate tuning can outperform many so-called state-of-the-art GNN models.

**Comment by willy-b** [comment]
Hey @sangyx , I would be happy to collaborate while fixing the misreported "#Params".

**Since we are updating your OGB entry and collaborating, we may as well increase the reported score and rank with a tiny change.**

I have proposed https://github.com/sangyx/gtrick/pull/11 which adds optional support for weight_decay, choice of add pooling instead of only mean pooling for converting node embeddings to graph embeddings, and option for a MLP after pooling node embeddings. These are off by default, controlled by argument to `gtrick/benchmark/pyg/graph_pred.py` only, so should not disrupt any existing usecases you have.

More details in the PR, **but I give example hyperparameters using these features (no hyperparameter search was done, I just used the ones I found in my own work, 2 layers, embed dimension of 64, 1e-6 weight decay, add pooling, use MLP after pooling node embeddings; though I did confirm with just 2 layers and 64 embed dimension no code changes the network in this project did not perform as well) that improve on your currently reported score** per the leaderboard methodology. See example sets of 10 runs here using your unmodified training/evaluation code to report the stats:  https://colab.research.google.com/drive/1MluOH3pOafwGdt9TKMqOBLfvLyeeDMX1?usp=sharing  (link public). **You would remain the smallest parameter count for a GIN on the leaderboard and the 2nd smallest model despite removing the invalid reported "#Params" of 7. The example I tested is 50,947 parameters. Still much smaller than 1.8M actual parameters in your current default configuration and in that of the OGB team's GIN entry (as we reduce layers from 5 to 2 and embedding dimension to 64 from 300).**

The reason to do this would be that as I had difficulty reproducing the previous ROCAUC score of this repo, **instead of potentially replacing your score with a slightly lower one when fixing the "#Params" reported, the PR above appears to consistently** (many sets of 10 runs, not seeing a single worse mean; though it is possible it rarely could do worse with new seeds) **outperform the existing reported score, so we could actually improve your ranking while fixing the parameter count.**

In doing this I am donating findings I was going to use for my own submission ( https://github.com/willy-b/tiny-GIN-for-ogbg-molhiv ) which I am happy to do as we don't need so many GINs on the leaderboard and I like the reusable structure of your project for enabling others to be more productive.

Let me know what you think of https://github.com/sangyx/gtrick/pull/11 , thanks!

**Comment by sangyx** [comment]
Hi @willy-b, thanks for your contribution. This is an excellent improvement. I've merged your PR and added a note to the readme.

**Comment by willy-b** [comment]
Thanks!

I am emailing OGB shortly to try to get the entry updated to correct the "#Params" column and update to latest score and will CC you.

**Comment by willy-b** [comment]
Hello,
As you know (but for the record here), I emailed them at 2024-07-18 23:12:50 -0700 (PT) notifying them you approved a correction of the parameter count in your solution from 7 to 50,947  (that is you accidentally reported the hyperparameter count on Apr 30, 2022 instead of parameter count) and **did not receive a follow up on that yet**:
```
Date: Thu, 18 Jul 2024 23:12:50 -0700
Subject: Fwd: Update to ogbg-molhiv entry #24 from Yunxin Sang (Per https://github.com/sangyx/gtrick/issues/10 )
From: Willy Bruns <adde.animulis@gmail.com>
To: ogb@cs.stanford.edu
Cc: Yunxin Sang <sangyunxin@gmail.com>
```
(note our emails are already visible at https://ogb.stanford.edu/docs/leader_graphprop/ )

I also mentioned it in another email to them about a different topic on 2024-08-23. They did take a couple of months to get back to me about something else so this may just be because the project is not very active and understaffed. The Google Group is essentially silent this year for example.

Thanks!
